crates/rpc/rpc/src/eth/api/fees.rs

```diff
@@ -100,110 +101,83 @@ where
+                fee_entries.push(entry);
+            }
+            // calculate rewards on the fly
         }
 
+        if fee_entries.len() != block_count as usize {
+            return Err(EthApiError::InvalidBlockRange)
+        }
         // Collect base fees, gas usage ratios and (optionally) reward percentile data
         let mut base_fee_per_gas: Vec<U256> = Vec::new();
         let mut gas_used_ratio: Vec<f64> = Vec::new();
         let mut rewards: Vec<Vec<U256>> = Vec::new();
-        for header in &headers {
-            base_fee_per_gas
-                .push(U256::try_from(header.base_fee_per_gas.unwrap_or_default()).unwrap());
-            gas_used_ratio.push(header.gas_used as f64 / header.gas_limit as f64);
 
-            // Percentiles were specified, so we need to collect reward percentile ino
+        for entry in &fee_entries {
+            base_fee_per_gas.push(U256::try_from(entry.base_fee_per_gas).unwrap());
+            gas_used_ratio.push(entry.gas_used_ratio);
+
             if let Some(percentiles) = &reward_percentiles {
-                rewards.push(self.calculate_reward_percentiles(percentiles, header).await?);
+                let mut block_rewards = Vec::new();
+
+                for &percentile in percentiles.iter() {
+                    block_rewards.push(self.fetch_reward_for_percentile(entry, percentile));
+                }
+                rewards.push(block_rewards);
             }
         }
 
         // The spec states that `base_fee_per_gas` "[..] includes the next block after the newest of
         // the returned range, because this value can be derived from the newest block"
         //
         // The unwrap is safe since we checked earlier that we got at least 1 header.
-        let last_header = headers.last().unwrap();
+
+        let last_entry = fee_entries.last().unwrap();
+
         let chain_spec = self.provider().chain_spec();
+
         base_fee_per_gas.push(U256::from(calculate_next_block_base_fee(
-            last_header.gas_used,
-            last_header.gas_limit,
-            last_header.base_fee_per_gas.unwrap_or_default(),
+            last_entry.gas_used,
+            last_entry.gas_limit,
+            last_entry.base_fee_per_gas,
             chain_spec.base_fee_params,
         )));
 
         Ok(FeeHistory {
             base_fee_per_gas,
             gas_used_ratio,
             oldest_block: U256::from(start_block),
-            reward: reward_percentiles.map(|_| rewards),
+            reward: if rewards.is_empty() { None } else { Some(rewards) },
         })
     }
 
-    /// Calculates reward percentiles for transactions in a block header.
-    /// Given a list of percentiles and a sealed block header, this function computes
-    /// the corresponding rewards for the transactions at each percentile.
-    ///
-    /// The results are returned as a vector of U256 values.
-    async fn calculate_reward_percentiles(
+    fn fetch_reward_for_percentile(
```

Comment:
please rename, nothing is being fetched here

this needs docs that this approximates with the configured resultion